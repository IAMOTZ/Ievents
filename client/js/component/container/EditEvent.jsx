import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import getAllCenters from '../../actions/centerActions';

import addEventStyles from '../../../sass/addEvent.scss';
import { UserSideNav } from '../common/SideNavigation.jsx';
import { UserTopNav } from '../common/TopNavigation.jsx';
import CenterOptions from '../common/CenterDropDown.jsx';
import Header from '../common/Header.jsx';
import { WarningAlert } from '../common/Alert';
import { updateEvent, clearStatus } from '../../actions/eventActions';

@connect((store) => {
  return { 
    user: store.user.user,
    centers: store.centers.centers,
    toEdit: store.events.toEdit,
    status: {
      error: store.events.status.updatingError.message,
      success: store.events.status.updated,
    }
  }
})

export default class EditEvent extends React.Component {
  constructor() {
    super();
    this.state = {
      title: null,
      description: null,
      date: null,
      centerId: null,
    };
  }

  componentDidMount() {
    this.props.dispatch(getAllCenters());
  }

  componentWillUnmount() {
    this.props.dispatch(clearStatus());
  }


  // This method uses user input to update the state
  getInput = (e) => {
    const state = this.state;
    state[e.target.name] = e.target.value;
    this.setState(state);
  }

  // This method fires the action to create an event
  update = () => {
    const {
     title,
      description,
      centerId
    } = this.state;
    let date = this.state.date ? this.state.date.replace(/-/g, '/') : null;
    const eventDetails = { title, description, date, centerId };
    const eventId = this.props.toEdit.id;
    this.props.dispatch(updateEvent(eventId, eventDetails, this.props.user.token));
  }

  render() {

    if (this.props.status.success) {
      return (<Redirect to="/events" />);
    } else {
      return (
        <div>
          {/* Top navigation on small screen */}
          <UserTopNav name={this.props.user.name} title='Add Event' />

          <div class="container-fluid">
            <div class="row">

              {/*  Side navigation on large screen */}
              <UserSideNav userName={this.props.user.name} />

              {/* Main content */}
              <div class="col-lg-10 offset-md-2" id="add-event-section">

                {/* Content Header(navigation) on large screen */}
                <Header text='Add Event' />

                {/* Input form */}
                <form class="mt-lg-5 w-lg-50">
                  <WarningAlert message={this.props.status.error} />
                  <div class="form-group">
                    <label for="title">Title</label>
                    <input type="text" class="form-control"
                      id="title" defaultValue={this.props.toEdit.title}
                      placeholder="A short description of your event"
                      name="title" onChange={this.getInput} />
                    <small id="emailHelp"
                      class="form-text text-muted">Less than 30 characters</small>
                  </div>
                  <div class="form-group">
                    <label for="description">Description</label>
                    <textarea class="form-control" id="description"
                      rows="6" defaultValue={this.props.toEdit.description}
                      placeholder="More details about the event"
                      name="description" onChange={this.getInput}></textarea>
                    <small id="emailHelp"
                      class="form-text text-muted">Less than 200 characters</small>
                  </div>
                </form>
                <form class="my-3 form-inline">
                  <div class="form-group">
                    <label for="date">Date</label>
                    <input type="date" id="date"
                      class="form-control mx-sm-3"
                      defaultValue={this.props.toEdit.date.replace(/\//g, '-')}
                      name="date" onChange={this.getInput} />
                  </div>
                  <div class="form-group">
                    <label for="centers">Choose a Center</label>
                    <select id="centers" class="form-control ml-md-3"
                      onChange={this.getInput}>
                      <option>choose a center</option>
                      <CenterOptions centers={this.props.centers} />
                    </select>
                  </div>
                </form>

                <a class="btn btn-outline-dark" onClick={this.update}>Update</a>
              </div>
            </div>
          </div>

          <footer class="d-block d-sm-none mt-5">
            <div class="container text-white text-center py-5">
              <h1>Ievents</h1>
              <p>Copyright &copy; 2017</p>
            </div>
          </footer>
        </div>
      );
    }
  }
}
