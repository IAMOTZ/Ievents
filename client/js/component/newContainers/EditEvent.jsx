import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { find } from 'lodash';
import { getAllCenters } from '../../actions/centerActions';
import { updateEvent } from '../../actions/eventActions';
import { stopAsyncProcess } from '../../actions/commonActions';
import * as asyncProcess from '../../actions/asyncProcess';
import { validateUpdateEventInputs } from '../../helpers/inputValidators';
import UserSideNav from '../common/SideNavigation';
import CenterOptions from '../common/CenterDropDown';
import Header from '../common/Header';
import { UserTopNav } from '../common/TopNavigation';
import { LoadingIcon } from '../common/LoadingAnimation';
import { BigAlert, SmallAlert } from '../common/Alert';

@connect((store) => {
  const { user } = store.authReducer;
  const { events } = store.fetchEventsReducer;
  return {
    user,
    isAdmin: user.role === 'admin',
    isSuperAdmin: user.role === 'admin' || user.role === 'superAdmin',
    isUserAuthenticaed: store.authReducer.loggingUserResolved,
    toUpdate: find(events, { id: store.updateEventReducer.eventToUpdate }),
    updatingEventStarted: store.updateEventReducer.updatingEventStarted,
    updatingEventResolved: store.updateEventReducer.updatingEventResolved,
    updatingEventError: store.updateEventReducer.updatingEventError,
    centers: store.fetchCentersReducer.centers,
  };
})
class EditEvent extends React.Component {
  constructor() {
    super();
    this.state = {
      title: null,
      description: null,
      date: null,
      centerId: null,
      inputErrors: {
        titleError: null,
        descriptionError: null,
        dateError: null,
        centerIdError: null,
      },
    };
  }

  componentDidMount() {
    this.props.dispatch(getAllCenters());
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.UPDATING_EVENT));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Clears all the inputErros in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      titleError: null,
      descriptionError: null,
      dateError: null,
      centerIdError: null,
    };
    this.setState(state);
  }

  /**
   * Dispatches the action to add the event.
   */
  update = () => {
    this.props.dispatch(stopAsyncProcess(asyncProcess.UPDATING_EVENT));
    const {
      title, description, centerId,
    } = this.state;
    const date = this.state.date ? this.state.date.replace(/-/g, '/') : null;
    const eventDetails = {
      title, description, date, centerId,
    };
    const inputErrors = validateUpdateEventInputs(eventDetails);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      const eventId = this.props.toUpdate.id;
      this.props.dispatch(updateEvent(eventId, eventDetails, this.props.user.token));
      window.scrollTo(0, 0);
    }
  }

  render() {
    if (!this.props.isUserAuthenticaed) {
      return (<Redirect to="/users/login" />);
    } else if (this.props.updatingEventResolved) {
      return (<Redirect to="/events" />);
    }
    return (
      <div className="add-event-container">
        {/* Top navigation on small screen */}
        <UserTopNav
          name={this.props.user.name}
          title="Edit Event"
          isAdmin={this.props.isAdmin}
          isSuperAdmin={this.props.isSuperAdmin}
          dispatch={this.props.dispatch}
        />
        <div className="container-fluid">
          <div className="row">
            {/*  Side navigation on large screen */}
            <UserSideNav
              name={this.props.user.name}
              isAdmin={this.props.isAdmin}
              isSuperAdmin={this.props.isSuperAdmin}
              dispatch={this.props.dispatch}
            />
            {/* Main content */}
            <div className="col-lg-10 offset-md-2" id="add-event-section">
              {/* Content Header(navigation) on large screen */}
              <Header text="Edit Event" />
              {/* Input form */}
              <form className="mt-lg-5 w-lg-50">
                <LoadingIcon start={this.props.updatingEventStarted} size={3} />
                <BigAlert message={this.props.updatingEventError} />
                <div className="form-group">
                  <label htmlFor="title">Title</label>
                  <input
                    type="text"
                    className="form-control"
                    id="title"
                    defaultValue={this.props.toUpdate.title}
                    placeholder="A short description of your event"
                    name="title"
                    onChange={this.getInput} />
                  <small
                    id="emailHelp"
                    className="form-text text-muted"
                  >Between 5 and 30 characters
                    </small>
                  <SmallAlert message={this.state.inputErrors.titleError} />
                </div>
                <div className="form-group">
                  <label htmlFor="description">Description</label>
                  <textarea
                    id="description"
                    rows="6"
                    className="form-control"
                    defaultValue={this.props.toUpdate.description}
                    placeholder="More details about the event"
                    name="description"
                    onChange={this.getInput} />
                  <small
                    id="emailHelp"
                    className="form-text text-muted"
                  >Less than 200 characters
                    </small>
                  <SmallAlert message={this.state.inputErrors.descriptionError} />
                </div>
              </form>
              <form className="my-3 form-inline">
                <div className="form-group">
                  <label htmlFor="date">Date</label>
                  <input
                    id="date"
                    type="date"
                    className="form-control ml-sm-3"
                    defaultValue={this.props.toUpdate.date.replace(/\//g, '-')}
                    name="date"
                    onChange={this.getInput}
                  />
                  <SmallAlert message={this.state.inputErrors.dateError} />
                </div>
                <div className="form-group ml-md-3">
                  <label htmlFor="centers">Choose a Center</label>
                  <select
                    id="centers"
                    className="form-control ml-md-3"
                    name="centerId"
                    onChange={this.getInput}
                    defaultValue={this.props.toUpdate.centerId}>
                    <option
                      value=""
                      name="centerId"
                    >choose a center
                      </option>
                    <CenterOptions centers={this.props.centers} />
                  </select>
                  <SmallAlert message={this.state.inputErrors.centerIdError} />
                </div>
              </form>
              <button
                className="btn btn-outline-dark"
                disabled={this.props.updatingEventStarted}
                onClick={this.update}
              >Update
                </button>
            </div>
          </div>
        </div>
        <footer className="d-block d-sm-none mt-5">
          <div className="container text-white text-center py-5">
            <h1>Ievents</h1>
            <p>Copyright &copy; 2017</p>
          </div>
        </footer>
      </div>
    );
  }
}

export default EditEvent;
