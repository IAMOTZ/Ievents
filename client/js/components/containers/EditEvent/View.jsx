import React from 'react';
import PropTypes from 'prop-types';
import EventForm from '../AddEvent/EventForm';
import SideNavigation from '../../common/SideNavigation';
import Header from '../../common/Header';
import { AuthTopNavigation } from '../../common/TopNavigation';

const View = props => (
  <div className="add-event-container">
    <AuthTopNavigation
      name={props.userName}
      title="Edit Event"
      isAdmin={props.isAdmin}
      isSuperAdmin={props.isSuperAdmin}
      dispatch={props.dispatch}
    />
    <div className="container-fluid">
      <div className="row">
        <SideNavigation
          name={props.userName}
          isAdmin={props.isAdmin}
          isSuperAdmin={props.isSuperAdmin}
          dispatch={props.dispatch}
        />
        <div className="col-lg-10 offset-md-2" id="main-content">
          <Header text="Edit Event" />
          <EventForm
            updatingEventStarted={props.updatingEventStarted}
            updatingEventError={props.updatingEventError}
            getInput={props.getInput}
            inputErrors={props.inputErrors}
            centers={props.centers}
            update={props.update}
            toUpdate={props.toUpdate}
          />
        </div>
      </div>
    </div>
    <footer className="d-block d-sm-none mt-5">
      <div className="container text-white text-center py-5">
        <h1>Ievents</h1>
        <p>Copyright &copy; 2017</p>
      </div>
    </footer>
  </div>
);

View.defaultProps = {
  inputErrors: {},
  updatingEventError: '',
};

/* eslint-disable react/forbid-prop-types */
View.propTypes = {
  userName: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  isSuperAdmin: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
  updatingEventStarted: PropTypes.bool.isRequired,
  updatingEventError: PropTypes.string,
  getInput: PropTypes.func.isRequired,
  inputErrors: PropTypes.object,
  centers: PropTypes.array.isRequired,
  update: PropTypes.func.isRequired,
  toUpdate: PropTypes.object.isRequired,
};

export default View;
