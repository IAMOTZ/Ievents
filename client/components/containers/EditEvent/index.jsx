import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { find } from 'lodash';
import { updateEvent } from '../../../actions/eventActions';
import { stopAsyncProcess } from '../../../actions/commonActions';
import * as asyncProcess from '../../../actions/asyncProcess';
import { validateUpdateEventInputs } from '../../../helpers/inputValidators';
import View from '../AddEvent/View';

@connect((store) => {
  const { user } = store.authReducer;
  const { events } = store.fetchEventsReducer;
  return {
    userName: user.name,
    userToken: user.token,
    isAdmin: user.role === 'admin' || user.role === 'superAdmin',
    isSuperAdmin: user.role === 'superAdmin',
    eventToUpdate: find(events, { id: store.updateEventReducer.eventToUpdate }),
    updatingEventStarted: store.updateEventReducer.updatingEventStarted,
    updatingEventResolved: store.updateEventReducer.updatingEventResolved,
    updatingEventError: store.updateEventReducer.updatingEventError,
  };
})
class EditEvent extends React.Component {
  constructor() {
    super();
    this.state = {
      title: null,
      description: null,
      date: null,
      centerId: null,
      inputErrors: {
        titleError: null,
        descriptionError: null,
        dateError: null,
      },
    };
  }

  componentWillUnmount() {
    this.props.dispatch(stopAsyncProcess(asyncProcess.UPDATING_EVENT));
  }

  /**
   * Stores the user inputs in the state of this component.
   * @param {Event} event The event object.
   */
  getInput = (event) => {
    const state = { ...this.state };
    state[event.target.name] = event.target.value;
    this.setState(state);
  }

  /**
   * Store the date selected in the state of this component.
   * @param {String} dateString The date in its string form.
   */
  handleDateSelection = (dateString) => {
    const date = new Date(dateString);
    const formattedDate = `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()}`;
    const state = { ...this.state };
    state.date = formattedDate;
    this.setState(state);
  }

  /**
   * Clears all the inputErros in the state.
   */
  clearInputErrors = () => {
    const state = { ...this.state };
    state.inputErrors = {
      titleError: null, descriptionError: null, dateError: null,
    };
    this.setState(state);
  }

  /**
   * Dispatches the action to add the event.
   */
  update = () => {
    this.props.dispatch(stopAsyncProcess(asyncProcess.UPDATING_EVENT));
    const { title, description, date } = this.state;
    const { centerId } = this.props.eventToUpdate;
    const eventDetails = {
      title, description, date, centerId,
    };
    const inputErrors = validateUpdateEventInputs(eventDetails);
    if (inputErrors.errorFound) {
      const state = { ...this.state };
      state.inputErrors = inputErrors;
      this.setState(state);
    } else {
      this.clearInputErrors();
      const eventId = this.props.eventToUpdate.id;
      this.props.dispatch(updateEvent(eventId, eventDetails, this.props.userToken));
      window.scrollTo(0, 0);
    }
  }

  render() {
    if (this.props.updatingEventResolved || !this.props.eventToUpdate) {
      return <Redirect to="/events" />;
    }
    return (
      <View
        userName={this.props.userName}
        isAdmin={this.props.isAdmin}
        isSuperAdmin={this.props.isSuperAdmin}
        dispatch={this.props.dispatch}
        getInput={this.getInput}
        inputErrors={this.state.inputErrors}
        update={this.update}
        updatingEventStarted={this.props.updatingEventStarted}
        updatingEventError={this.props.updatingEventError}
        eventToUpdate={this.props.eventToUpdate}
        updating
        handleDateSelection={this.handleDateSelection}
        selectedDate={this.state.date || this.props.eventToUpdate.date}
      />
    );
  }
}

EditEvent.defaultProps = {
  userName: '',
  userToken: '',
  isAdmin: false,
  isSuperAdmin: false,
  eventToUpdate: {},
  updatingEventStarted: false,
  updatingEventResolved: false,
  updatingEventError: '',
  dispatch: () => { },
};

/* eslint-disable react/forbid-prop-types */
EditEvent.propTypes = {
  userName: PropTypes.string,
  userToken: PropTypes.string,
  isAdmin: PropTypes.bool,
  isSuperAdmin: PropTypes.bool,
  eventToUpdate: PropTypes.object,
  updatingEventStarted: PropTypes.bool,
  updatingEventResolved: PropTypes.bool,
  updatingEventError: PropTypes.string,
  dispatch: PropTypes.func,
};

export default EditEvent;
