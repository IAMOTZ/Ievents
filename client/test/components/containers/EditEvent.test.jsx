import React from 'react';
import EditEvent from '../../../components/containers/EditEvent';

describe('<EditEvent />', () => {
  const props = {
    userName: 'tester',
    userToken: 'justASampleToken',
    isAdmin: false,
    isSuperAdmin: false,
    eventToUpdate: { id: 5 },
    updatingEventStarted: false,
    updatingEventResolved: false,
    updatingEventError: null,
    dispatch: jest.fn()
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering:', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<EditEvent {...props} />);
    });
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should redirect to the events page if updating event resolved', () => {
      wrapper.setProps(alterProps({ updatingEventResolved: true }));
      expect(wrapper).toMatchSnapshot();
    });
    it('should redirect to the events page if there is no event to update', () => {
      wrapper.setProps(alterProps({ eventToUpdate: null }));
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Behaviour:', () => {
    let wrapper; let wrapperInstance;
    beforeEach(() => {
      props.dispatch.mockReset();
      wrapper = shallow(<EditEvent {...props} />);
      wrapperInstance = wrapper.instance();
    });
    // The names on the following test block refers
    // to the instance method that handles the behaviour tested.
    describe('GET_INPUT', () => {
      it('should update the state with user input', () => {
        const userInput = 'edited event';
        const fakeEvent = { target: { name: 'title', value: userInput } };
        wrapperInstance.getInput(fakeEvent);
        expect(wrapper.state('title')).toEqual(userInput);
      });
    });

    describe('CLEAR_INPUT_ERRORS', () => {
      it('should clear input errors in the state', () => {
        const titleError = 'Event title is required';
        wrapper.setState({ inputErrors: { titleError } });
        wrapperInstance.clearInputErrors();
        expect(wrapper.state('inputErrors').titleError).toBeNull();
      });
    });

    describe('HANDLE_DATE_SELECTION', () => {
      it('should update the state with date input being formatted', () => {
        const date = new Date();
        const dateString = date.toDateString();
        const formattedDate = `${date.getFullYear()}/${date.getMonth() + 1}/${date.getDate()}`;
        wrapperInstance.handleDateSelection(dateString);
        expect(wrapper.state('date')).toEqual(formattedDate);
      });
    });

    describe('UPDATE', () => {
      it('should dispatch an action and set inputErros if validation fails', () => {
        const emptyTitle = '';
        wrapper.setState({ title: emptyTitle });
        props.dispatch.mockReset();
        wrapperInstance.update();
        expect(typeof wrapper.state('inputErrors').titleError).toEqual('string');
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
      it('should dispatch two acitons and clear inputErrors if validation pass', () => {
        const correctEventData = {
          title: 'test event',
          description: 'this is the test event',
          date: '2018/06/10',
        };
        const previousError = 'Title is required';
        wrapper.setState({
          ...correctEventData, inputErrors: { titleError: previousError }
        });
        props.dispatch.mockReset();
        wrapperInstance.update();
        expect(wrapper.state('inputErrors').titleError).toBeNull();
        expect(props.dispatch.mock.calls.length).toEqual(2);
      });
    });

    describe('COMOPONENT_WILL_UNMOUNT', () => {
      it('should dispatch an action', () => {
        props.dispatch.mockReset();
        wrapper.unmount();
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
    });
  });
});
