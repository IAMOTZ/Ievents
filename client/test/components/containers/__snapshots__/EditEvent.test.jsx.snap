// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditEvent /> Rendering: should redirect to the events page if there is no event to update 1`] = `
<Redirect
  push={false}
  to="/events"
/>
`;

exports[`<EditEvent /> Rendering: should redirect to the events page if updating event resolved 1`] = `
<Redirect
  push={false}
  to="/events"
/>
`;

exports[`<EditEvent /> Rendering: should render correctly 1`] = `
<View
  add={[Function]}
  addingEventError=""
  addingEventStarted={false}
  centerToBook={Object {}}
  dispatch={[MockFunction]}
  eventToUpdate={
    Object {
      "id": 5,
    }
  }
  getInput={[Function]}
  handleDateSelection={[Function]}
  inputErrors={
    Object {
      "dateError": null,
      "descriptionError": null,
      "titleError": null,
    }
  }
  isAdmin={false}
  isSuperAdmin={false}
  update={[Function]}
  updating={true}
  updatingEventError={null}
  updatingEventStarted={false}
  userName="tester"
/>
`;
